### ISSUE TITLE
Rewhere on a has_many association drops WHERE clause in 4.2.4

### ISSUE THREAD

**Issue by scottohara** [rails/rails#21955]
In 4.2.4, it appears that using a `rewhere` scope on a `has_many` relation is broken  (or at least, has materially changed from earlier versions).

The following gists demonstrate the issue:

### activerecord 4.2.3  ('working' example)
https://gist.github.com/scottohara/08d73351e27adeb6f694

In this example, a `rewhere` scope on the `has_many :comments` association arbitrarily overrides the `WHERE` clause to a different id  (ie. `-> { rewhere(post_id: 99) }`).

The SQL `SELECT` statement for `post.comments` is as expected:

```
SELECT "comments".* FROM "comments" WHERE "comments"."post_id" = 99
```

### activerecord 4.2.4 ('broken' example)
https://gist.github.com/scottohara/70a07172882e067aaa96

This is identical to the above gist, but with the activerecord gem version bumped to 4.2.4.

The presence of the `rewhere` scope now causes the SQL `SELECT` statement to lose its `WHERE` clause entirely; yielding:

```
SELECT "comments".* FROM "comments"
```

The same behaviour (as 4.2.4) is present in edge (master).

Is this a regression in 4.2.4, or an expected behaviour change?  (and if expected, what would be the correct way to use `rewhere` in 4.2.4?)

**Comment by jbranchaud** [comment]
It looks like `rewhere` uses `unscope(...).where(...)` under the hood. I think the `where` constraints are not being applied after the unscoping.

**Comment by rafaelfranca** [comment]
Fixed in 2ba92bb8090b676b664c793f45cbe40e51cb1832 and backported in 1cfcf414a5919ec3fd9e6fb01bfaef62e5ff46e5
